### PR TITLE
LTD-3389-3390: Model changes to record Countersigning data

### PR DESCRIPTION
**Pull request by saruniitr**
## Change description

For LU countersigning we need to record countersigning decision also as user can accept/reject original outcome.
The current countersigning fields in `Advice` model doesn't support this.
We can extend the above fields but for some cases double countersigning is required.
To support all these requirements and future proof it a separate model is introduced to save countersigning data.
This will support multiple countersigners and also give us the flexibility to add additional fields.

LU countersigning work is under a feature flag so to better isolate UI changes without affecting countersigning flow for other teams (eg FCDO) a new endpoint is also created that uses this new model. FCDO can continue to use existing flow and eventually they will be migrated to use this in future.

Add tests to cover possible cases.

### CODE REVIEW

File: api/cases/models.py [hunk 1]
```diff
@@ -537,6 +537,16 @@ def equals(self, other):
         )
 
 
+class Countersign(TimestampableModel):
+    order = models.PositiveIntegerField(help_text="Indicates countersigning order")
+    outcome_accepted = models.BooleanField(default=None)
+    reasons = models.TextField(
```
Comment (saruniitr) [inline api/cases/models.py:543]: User need to provide reasons whether they accept/reject recommendation. It is mandatory in both cases.
The plan is to use a single field but open to changing to separate fields also.

File: api/applications/serializers/advice.py [hunk 2]
```diff
@@ -196,6 +197,23 @@ class Meta:
...
+        return result
+
+
+class CountersignAdviceWithDecisionSerializer(serializers.ModelSerializer):
+    countersigned_user = serializers.PrimaryKeyRelatedField(queryset=GovUser.objects.filter(status=UserStatuses.ACTIVE))
+    advice = serializers.PrimaryKeyRelatedField(queryset=Advice.objects.all())
+
+    class Meta:
+        model = Countersign
+        fields = "__all__"
```
Comment (kevincarrogan) [inline api/applications/serializers/advice.py:213]: Isn't this considered an anti-pattern and we should specify fields explicitly?

File: api/applications/serializers/advice.py [hunk 3]
```diff
@@ -196,6 +197,23 @@ class Meta:
         list_serializer_class = CountersignAdviceListSerializer
 
 
+class CountersignAdviceWithDecisionListSerializer(serializers.ListSerializer):
+    def update(self, instances, validated_data):
+        instance_map = {index: instance for index, instance in enumerate(instances)}
+        result = [self.child.update(instance_map[index], data) for index, data in enumerate(validated_data)]
+        return result
```
Comment (kevincarrogan) [inline api/applications/serializers/advice.py:204]: What's the use case for us updating multiple decisions at once?

File: api/cases/models.py (hunk 1 above)
Comment (kevincarrogan) [reply api/cases/models.py:543]: Once we create this object will they ever change the outcome?

File: api/cases/views/views.py [hunk 4]
```diff
@@ -1040,6 +1046,84 @@ def put(self, request, **kwargs):
...
+            department = "department"
+
+        audit_trail_service.create(
+            actor=request.user,
+            verb=AuditType.COUNTERSIGN_ADVICE,
+            target=case,
+            payload={"department": department},
+        )
+
+        return JsonResponse({"countersigned_advice": serializer.data}, status=status.HTTP_200_OK)
```
Comment (kevincarrogan) [inline api/cases/views/views.py:1124]: There appears to be a lot of overlap here with the `post` method. Could this be consolidated in some way?

File: api/cases/models.py (hunk 1 above)
Comment (saruniitr) [reply api/cases/models.py:543]: Before moving case foward they will have the option to Edit their countersigning decision.
So initially they may have said "Yes" but they could change their mind to "No" and then move case forward.

File: api/applications/serializers/advice.py (hunk 3 above)
Comment (saruniitr) [reply api/applications/serializers/advice.py:204]: explained in internal channel

**Review by antroy-madetech — COMMENTED** [review]
Presumably we'll do the migration to copy the existing text field countersignatures into the new table when we do the work to move all other countersignatures to the new model?

**Review by antroy-madetech — COMMENTED** [review]
Looks like codecov has picked up a few lines of code that are missing tests. Would be good to get those covered.

File: api/cases/views/views.py [hunk 5]
```diff
@@ -1040,6 +1046,84 @@ def put(self, request, **kwargs):
...
+            department = department.name
+        else:
+            department = "department"
+
+        audit_trail_service.create(
+            actor=request.user,
+            verb=AuditType.COUNTERSIGN_ADVICE,
+            target=case,
+            payload={"department": department},
+        )
```
[the remark this answers is not in the document]
Comment (saruniitr) [reply api/cases/views/views.py:1122]: :+1:

File: api/cases/views/views.py [hunk 6]
```diff
@@ -1040,6 +1046,71 @@ def put(self, request, **kwargs):
         return JsonResponse({"advice": serializer.data}, status=status.HTTP_200_OK)
 
 
+class CountersignAdviceV2(APIView):
```
Comment (depsiatwal) [inline api/cases/views/views.py:1049]: V2 ? 
can we get a better name ?
are migrating off original one

Comment (saruniitr) [reply api/cases/views/views.py:1049]: @depsiatwal 
yeah open to suggestions.

The main difference is countersigners will be able to provide accept/refuse decision now so I included "WithDecision" for some of the serializer names so may similar can be done here?
Please suggest any other alternatives.

**Review by antroy-madetech — APPROVED** [review]

File: api/cases/views/views.py (hunk 6 above)
Comment (depsiatwal) [reply api/cases/views/views.py:1049]: perhaps we can use CountersignDecisionAdvice

File: api/applications/serializers/advice.py [hunk 7]
```diff
@@ -196,6 +197,33 @@ class Meta:
...
+    case = serializers.PrimaryKeyRelatedField(queryset=Case.objects.all())
+    advice = serializers.PrimaryKeyRelatedField(queryset=Advice.objects.all())
+
+    class Meta:
+        model = CountersignedAdvice
+        fields = ("id", "order", "outcome_accepted", "reasons", "countersigned_user", "case", "advice")
+        list_serializer_class = CountersignAdviceWithDecisionListSerializer
+
+
+class CountersignedAdviceWithDecisionViewSerializer(serializers.Serializer):
```
Comment (hnryjmes) [inline api/applications/serializers/advice.py:218]: Please can this be `CountersignAdviceWithDecisionViewSerializer` (Countersigned -> Countersign) if possible

File: api/cases/models.py [hunk 8]
```diff
@@ -537,6 +537,18 @@ def equals(self, other):
         )
 
 
+class CountersignedAdvice(TimestampableModel):
```
Comment (hnryjmes) [inline api/cases/models.py:540]: Please can this be `CountersignAdvice` (Countersigned -> Countersign) if possible

File: api/cases/views/views.py [hunk 9]
```diff
@@ -1040,6 +1046,71 @@ def put(self, request, **kwargs):
...
+            CountersignedAdvice.objects.filter(id__in=countersign_ids), data=request.data, partial=True, many=True
+        )
+        if not serializer.is_valid():
+            return JsonResponse({"errors": serializer.errors}, status=status.HTTP_400_BAD_REQUEST)
+
+        serializer.save()
+
+        self.audit_countersign(request, case)
+
+        return JsonResponse({"countersigned_advice": serializer.data}, status=status.HTTP_200_OK)
```
Comment (hnryjmes) [inline api/cases/views/views.py:1111]: Please can this be `{"countersign_advice": serializer.data}` (countersigned -> countersign) if possible

File: api/cases/models.py (hunk 8 above)
Comment (saruniitr) [reply api/cases/models.py:540]: ok :+1:

File: api/cases/views/views.py [hunk 10]
```diff
@@ -1040,6 +1046,71 @@ def put(self, request, **kwargs):
...
+
+        serializer = self.serializer_class(data=data, many=True)
+        if not serializer.is_valid():
+            return JsonResponse({"errors": serializer.errors}, status=status.HTTP_400_BAD_REQUEST)
+
+        serializer.save()
+
+        self.audit_countersign(request, case)
+
+        return JsonResponse({"countersigned_advice": serializer.data}, status=status.HTTP_201_CREATED)
```
Comment (hnryjmes) [inline api/cases/views/views.py:1095]: Please can this be `{"countersign_advice": serializer.data}` (countersigned -> countersign) if possible

File: api/cases/tests/test_get_case.py [hunk 11]
```diff
@@ -131,6 +132,21 @@ def test_case_returns_has_advice(self):
         self.assertIn("my_team", has_advice_response_data)
         self.assertIn("final", has_advice_response_data)
 
+    def test_case_returns_countersigned_advice(self):
+        case = self.submit_application(self.standard_application)
+        url = reverse("cases:case", kwargs={"pk": case.id})
+        advice = FinalAdviceFactory(case=case, user=self.gov_user)
+        countersigned_advice = CountersignAdviceFactory(case=case, advice=advice)
```
Comment (hnryjmes) [inline api/cases/tests/test_get_case.py:139]: Please can this be `countersign_advice` (countersigned -> countersign) if possible

File: api/applications/serializers/advice.py (hunk 7 above)
Comment (saruniitr) [reply api/applications/serializers/advice.py:218]: Actually this is already named as `Countersign...`
but following @depsiatwal suggestion above, I will update this as `CountersignDecisionAdviceViewSerializer`

File: api/cases/views/views.py (hunk 9 above)
Comment (saruniitr) [reply api/cases/views/views.py:1111]: :+1:

File: api/cases/views/views.py (hunk 10 above)
Comment (saruniitr) [reply api/cases/views/views.py:1095]: :+1:

File: api/cases/tests/test_get_case.py (hunk 11 above)
Comment (saruniitr) [reply api/cases/tests/test_get_case.py:139]: :+1:

**Comment by hnryjmes** [comment]
I think you already understand this @saruniitr  but for the avoidance of doubt, yes, everywhere we have `countersigned` as a variable name I think this should be `countersign` just because I feel strongly we should just pick one and make it consistent everywhere, even if 'countersigned' might sometimes be more appropriate in written English. I know this is a lot to change but I think future devs will appreciate it.

File: api/cases/models.py [hunk 12]
```diff
@@ -537,6 +537,18 @@ def equals(self, other):
...
+    id = models.UUIDField(primary_key=True, default=uuid.uuid4, editable=False)
+    order = models.PositiveIntegerField(help_text="Indicates countersigning order")
+    outcome_accepted = models.BooleanField()
+    reasons = models.TextField(
+        help_text="Reasons provided by the countersigner when they agree/disagree with the advice during countersigning",
+    )
+    countersigned_user = models.ForeignKey(GovUser, on_delete=models.DO_NOTHING, related_name="countersigned_user")
```
Comment (hnryjmes) [inline api/cases/models.py:547]: not really a suggestion, more of a question for my own understanding - if we have `on_delete=models.DO_NOTHING` shown here, that means, when a GovUser is deleted, this field then contains a reference to a deleted object, or the field contains None ?

Comment (saruniitr) [reply api/cases/models.py:547]: I think it becomes `None`. We want to keep this object in that case for auditing purposes.

File: api/cases/tests/test_create_advice.py [hunk 13]
```diff
@@ -267,3 +271,151 @@ def test_countersign_advice_success(self):
         audit_qs = Audit.objects.filter(verb=AuditType.COUNTERSIGN_ADVICE)
         self.assertEqual(audit_qs.count(), 1)
         self.assertEqual(audit_qs.first().actor, self.gov_user)
+
+
+class CountersignAdviceWithDecisionTests(DataTestClient):
```
Comment (depsiatwal) [inline api/cases/tests/test_create_advice.py:276]: does this now require renaming remove 'with'

**Review by hnryjmes — APPROVED** [review]

**Pull request merged**